### PR TITLE
Fix PUPPI MET tails

### PR DESCRIPTION
**Pull request by ahinzmann**
This contains 2 fixes needed to remove tails of the PUPPI MET distribution:
- Photon ID was not applied in PUPPI Photon code
- PUPPI photon code was not protected against adding multiple copies of the same PF candidate in the output collected
The impact of this PR on jet and MET response and resolution has been validated.
Slide 17-22 show the impact of this change to MET:
https://indico.cern.ch/event/587258/contributions/2366495/attachments/1368938/2117761/satoshi_updatedAfterTheMeeting.pdf
What is labeled v10 is in the CMSSW_8_0_>=20 and CMSSW_8_1_X and CMSSW_9_0_X releases at the moment.
What is labeled "PFPUPPI v10 PHfix wgtRecal" is in this PR.

This PR changes the MINIAOD PUPPI MET collection, no changes to RECO/AOD event content.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: CommonTools/PileupAlgos/plugins/PuppiPhoton.cc [hunk 1]
```diff
@@ -31,17 +31,19 @@ PuppiPhoton::PuppiPhoton(const edm::ParameterSet& iConfig) {
   tokenPFCandidates_     = consumes<CandidateView>(iConfig.getParameter<edm::InputTag>("candName"));
   tokenPuppiCandidates_  = consumes<CandidateView>(iConfig.getParameter<edm::InputTag>("puppiCandName"));
   tokenPhotonCandidates_ = consumes<CandidateView>(iConfig.getParameter<edm::InputTag>("photonName"));
-  tokenPhotonId_         = consumes<edm::ValueMap<bool>  >(iConfig.getParameter<edm::InputTag>("photonId")); 
+  reco2pf_               = mayConsume<edm::ValueMap<std::vector<reco::PFCandidateRef> > >(iConfig.getParameter<edm::InputTag>("recoToPFMap"));
```
Comment (slava77) [inline CommonTools/PileupAlgos/plugins/PuppiPhoton.cc]: please change to consumes according to the value of runOnMiniAOD.
The current implementation has no alternatives if runOnMiniAOD is false.

File: CommonTools/PileupAlgos/plugins/PuppiPhoton.cc [hunk 2]
```diff
@@ -31,17 +31,19 @@ PuppiPhoton::PuppiPhoton(const edm::ParameterSet& iConfig) {
...
+  tokenPhotonId_         = mayConsume<edm::ValueMap<bool>  >(iConfig.getParameter<edm::InputTag>("photonId"));
```
Comment (slava77) [inline CommonTools/PileupAlgos/plugins/PuppiPhoton.cc]: please change to consumes based on the value of usePhotonId.
Actually, if photonId is empty, the framework will already ignore the ```consumes```

File: CommonTools/PileupAlgos/plugins/PuppiPhoton.cc [hunk 3]
```diff
@@ -74,25 +79,32 @@ void PuppiPhoton::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
...
+      if(pElectron != 0) {
+        for( const edm::Ref<pat::PackedCandidateCollection> & ref : pElectron->associatedPackedPFCandidates() )
+	  if(fabs(pfCol->ptrAt(ref.key())->eta()) < eta_ )  {
+	    phoIndx.push_back(ref.key());
+	    phoCands.push_back(&(*(pfCol->ptrAt(ref.key()))));
+	  }
+      }
+    } else {
+      for( const edm::Ref<std::vector<reco::PFCandidate> > & ref : (*reco2pf)[phoCol->ptrAt(iC)] ) {
+	  if(fabs(pfCol->ptrAt(ref.key())->eta()) < eta_ )  {
```
Comment (slava77) [inline CommonTools/PileupAlgos/plugins/PuppiPhoton.cc]: here again it looks like ``` pfCol->ptrAt(ref.key()) ``` should be replaced with ```ref```, else this opens up a way for inconsistencies

**Comment by slava77** [comment]
Here are some comparisons on top of #17067 
The baseline (black) is 900pre2 + #17067   using 900pre2 RECO files as an inputs; the "new" is this PR + #17067  

HGG PU35, in the standard AOD->MINIAOD workflow

all_sign817a-patvssign816-pat_h125gg13tevpuwf25203p0c_patmets_slimmedmetspuppi__pat_obj_pt

since the genMET is essentially zero, the diff with genmet shows an increase in the over-measured values
wf25203_7avs6_puppimet_gendiff

e.g. the slice of MET 80 to 100 more than doubles in population
wf25203_7avs6_puppimet_gendiff_80to100

I'm guessing it's the ID cuts which are now applied
wf25203_7avs6_puppimet_photonetfrac



HGG no PU
all_sign817a-patvssign816-pat_h125gg13tevwf1332p0c_patmets_slimmedmetspuppi__pat_obj_pt

ZEE PU35 and TTbar PU35 do not have a significant change.

=============

so, this PR makes PUPPI MET worse for events with photons.
The increase in MET tail from this PR itself in events with photons is slightly worse in no-PU (the improvements in #17067  balance out the degradation in MET from this PR).

Looking at the fix referenced in the slides
https://github.com/cms-met/cmssw/compare/METRecipe_8020...ahinzmann:METRecipe_8020_Moriond17
I notice only the difference in ID for the code part.

The slides suggest that puppiMET is remade from miniAOD inputs.
It looks like we are again hitting the problem that the code is developed and debugged based on rerunning on top of miniAOD and it behaves differently compared to running on AOD (the default puppi and other miniAOD products). 
This kind of defeats the purpose of having puppi in standard miniAOD outputs.

Please make some plots comparing puppiMET made from AOD and that made from miniAOD.
If there is decent agreement, maybe the ID is to blame and it could be possibly updated to a more recent one.

**Comment by ahinzmann** [comment]
@slava77 I implemented the comments and did a comparison of running on AOD and MiniAOD. Starting from 1000 event in
/store/relval/CMSSW_9_0_0_pre2/RelValH125GGgluonfusion_13/GEN-SIM-RECO/PU25ns_90X_mcRun2_asymptotic_v0-v1/10000/167B6FB3-B4C2-E611-AAEE-0CC47A7C3410.root
I ran the MiniAOD sequence from cmsDriver.
Then reran PUPPI on top of this MiniAOD.
I also enabled "useExistingWeights" when rerunning PUPPI to minimize effects of numerical precision.
The resulting comparison is attached.
met

The MET value when running on AOD and MiniAOD agrees within 1% for all events. For example, the first couple of events differ like this:

< PUPPI MET: pt   8.1
---
> PUPPI MET: pt   8.2
258c258
< PUPPI MET: pt  23.5
---
> PUPPI MET: pt  23.1
261c261
< PUPPI MET: pt  14.0
---
> PUPPI MET: pt  14.1
264c264
< PUPPI MET: pt  28.4
---
> PUPPI MET: pt  28.5
267c267
< PUPPI MET: pt  53.5
---
> PUPPI MET: pt  53.7
285c285
< PUPPI MET: pt  33.5
---
> PUPPI MET: pt  33.3
288c288
< PUPPI MET: pt  15.9
---
> PUPPI MET: pt  15.8
291c291
< PUPPI MET: pt  11.7
---
> PUPPI MET: pt  11.8
300c300
< PUPPI MET: pt  47.4
---
> PUPPI MET: pt  47.5

So the behavior you observe does not come from AOD/MiniAOD differences. I also doubt it comes from the photon ID, since everything was tested with the Spring15 photon ID that is also in CMSSW_9_0_0. The change of the METRecipe_8020_Moriond17 recipe to the Spring16 photon ID only happened recently with the new Moriond recommendations.

Do you think you could check one more plot, to confirm that this PR achieves its main purpose, that is removing the MET tails in QCD events as shown on slide 3 of https://indico.cern.ch/event/587258/contributions/2366495/attachments/1368938/2117761/satoshi_updatedAfterTheMeeting.pdf
A plot of the MET in log-y-scale in a flat (or high pT) QCD sample should show this.
It could be possible that getting rid of these tails trades off with some performance loss in the gamma gamma MET.

**Comment by slava77** [comment]
On 1/6/17 1:57 PM, ahinzmann wrote:
> Do you think you could check one more plot, to confirm that this PR
> achieves its main purpose, that is removing the MET tails in QCD events
> as shown on slide 3 of
> https://indico.cern.ch/event/587258/contributions/2366495/attachments/1368938/2117761/satoshi_updatedAfterTheMeeting.pdf
> A plot of the MET in log-y-scale in a flat (or high pT) QCD sample
> should show this.
> It could be possible that getting rid of these tails trades off with
> some performance loss in the gamma gamma MET.

I can check in the QCD sample, but mostly likely not until Monday,
possibly during the weekend.

Just to be more specific on the sample:
which sample is used in p3?
The text says "flat QCD (MiniAOD one file)."
Is there PU in this sample?

p5 has
/QCD_Pt-15to7000_TuneCUETP8M1_Flat_13TeV_pythia8/RunIISummer16MiniAODv2-PUFlat0to70_magnetOn_80X_mcRun2_asymptotic_2016_TrancheIV_v4-v1/MINIAODSIM
with pileup
Was it this sample?

**Comment by ahinzmann** [comment]
@slava77 confirmed. The 0-70 PU sample. (Though PU is not the source of the tails, thus the MET tails should show up in any flat/high pT QCD samples)

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
Here is a diff for 25209.0 miniAOD (from RECO /RelValQCD_FlatPt_15_3000HS_13/CMSSW_9_0_0_pre2-PU25ns_90X_mcRun2_asymptotic_v0-v1/GEN-SIM-RECO )
900pre2 as a baseline (black) and this 900pre2+this PR cherry-picked (in red)

wf25209_metpuppi_20gevbins

clearly, there is a reduction of large tails.
This appears to come with some broadening of the core, which is slightly visible in the plot above as well as in the slide 3 from 
https://indico.cern.ch/event/587258/contributions/2366495/attachments/1368938/2117761/satoshi_updatedAfterTheMeeting.pdf
satoshi_p3_zoom

So, the Hgammagamma plots I posted earlier are also consistent.

@ahinzmann 
Based on the context and your confirmation of MET remade from miniAOD and from AOD, it looks like this can be signed off.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17072	acd025b
- code changes are in line with the description and follow up review
- jenkins test pass and comparisons show changes only in puppi-related variables
- results of tests with larger samples were posted earlier

**Pull request merged**